### PR TITLE
Move creation of mpm to optimize_final_module

### LINKED ISSUE #9646: config.OPT is not overridden with override_config('OPT', 0)
Hi, 

I'm trying to migrate Numba to LLVM's new pass manager APIs introduced in [1046](https://github.com/numba/llvmlite/pull/1046) and I think I might have encountered a bug.
Some of the tests (eg test_debuginfo.py) make use of this snippet to override default config flags
```
with override_config('OPT', 0):
    fn.compile(sig)
    return fn.inspect_llvm(sig)
```
For example, when we execute the test `test_debuginfo.py:test_DILocation` `core/codegen.py:_optimize_final_module` makes a call to `_mpm_cheap.run()` and `_mpm_full.run()` with `config.OPT` correctly overriden to 0 (as expected).

However these pass manager objects are already created and exist even before `_optimize_final_module` was called (core/codegen.py:line 1211) with default `config.OPT` levels, meaning `override_config` snippet didn't have any effect.

I might be missing something here and this might not actually be a bug but nonetheless I would like to propose a solution/code-change  to shift creating of these pass manager objects inside `_optimize_final_module` function where they are used. It would also solve the problem of using same pass manager object and calling run() on it multiple times.

### PR DESCRIPTION
**Pull request by yashssh**
Fixes #9646

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the patch @yashssh. Please could you add a release note to accompany this patch (instructions are here along with examples from other PRs in the directory above: https://github.com/numba/numba/blob/f3665c45463ce9ef561dfa10d6bff27079524862/docs/upcoming_changes/README.rst?plain=1#L1-L47)? I think this is an "improvement" type of change. Many thanks!

File: numba/tests/test_override.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        name = foo.overloads[foo.signatures[0]].fndesc.mangled_name
+        funcs = [x for x in module.functions if x.name == name]
+        self.assertEqual(len(funcs), 1)
+        func = funcs[0]
+        blocks = [x for x in func.blocks]
+        self.assertGreater(len(blocks), 1)
+        block = blocks[0]
+
+        # Find sequence without non-debug instructions
```
Comment (stuartarchibald) [inline numba/tests/test_override.py:41]:
```suggestion
        # Find sequence with non-debug instructions
```
I think this is looking for the point at which the arithmetic instructions start and is discounting `call` as those are associated with `llvm.dbg.declare`.

File: numba/tests/test_override.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        for x in instrs:
+            if x.opcode in op_expect:
+                op_expect.remove(x.opcode)
+                if not started:
+                    started = True
+            elif op_expect and started:
+                break
+
+        self.assertGreater(len(op_expect), 0,
+                           "Function was optimized unexpectedly")
```
Comment (stuartarchibald) [inline numba/tests/test_override.py:54]: I've verified this manually against the IR produced on `main` and the IR produced with this PR. The IR produced when using this PR contains `load`s and `store`s interleaved with the `fadd`, `fmul` and `fdiv` instructions, whereas on `main` these loads and stores are optimised out and the arithmetic instructions appear sequentially.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch and for investigating this @yashssh, there's a few minor things to resolve else looks good. It's great to see this issue fixed.

File: numba/tests/test_override.py
[the remark this answers is not in the document]
Comment (yashssh) [reply numba/tests/test_override.py:1]: Renamed it to test_optimisation_pipelines.py. Will add more tests to it for various optimisation levels when pushing code for NewPassManager changes.

**Comment by yashssh** [comment]
Thanks for the review @stuartarchibald! I have added the release notes and addressed the comments.

File: numba/tests/test_override.py
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/tests/test_override.py:1]: Great, thanks for doing that.

**Comment by stuartarchibald** [comment]
I'm just going to close and reopen this to trigger a rebuild on RTD now that #9675 resolved the issue.

**Comment by stuartarchibald** [comment]
@yashssh I think that this branch is going to need to have `numba:main` merged in to pick up the changes in #9675 so that RTD will build.

**Comment by yashssh** [comment]
Are you suggesting to rebase my branch on top of main?

**Comment by stuartarchibald** [comment]
> Are you suggesting to rebase my branch on top of main?

I suggest merging `HEAD` (currently 99f5cc6) of `numba:main` into this branch. If you do a rebase all the commit SHAs change and github sometimes struggles to track comments/review across such a change, which can then result in having to do a re-review of the entire PR. Thanks!

**Review by gmarkall — APPROVED** [review]
The merge from `main` looks good.

**Pull request merged**